src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,18 +31,25 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
+            if (provider == null)
+            {
+                throw new ArgumentNullException(nameof(provider));
+            }
+
             int bestLength = -1;
+            bool multipleBestLengthFound = false;
             bool seenPreferred = false;
 
             ConstructorMatcher bestMatcher = default;
+            IServiceProviderIsService? serviceProviderIsService = provider.GetService<IServiceProviderIsService>();
 
             if (!instanceType.IsAbstract)
             {
                 foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
                 {
                     var matcher = new ConstructorMatcher(constructor);
                     bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
+                    int length = matcher.Match(parameters, serviceProviderIsService);
```

Comment:
Maybe one thing that should be changed here is what @halter73 mentioned:

> I think there should at least be a way for developers to force ActivatorUtilities to use a specific constructor even if an IServiceProviderIsService exists and IServiceProviderIsService.IsService(parameterType) returns false

With the old code, if a ctor has `ActivatorUtilitiesConstructorAttribute` on it and it takes all the `parameters` supplied to `CreateInstance`, we will use that ctor.

With the new code, if a ctor has `ActivatorUtilitiesConstructorAttribute` on it and takes all the `parameters` supplied to `CreateInstance`, but the `IServiceProviderIsService` gives a false negative on one of the IsService calls for a different parameter, we will not use the ctor but throw an exception.

So maybe we need to pass `isPreferred` into `matcher.Match`, and only check the supplied `parameters` when `isPreferred=true`.

Reply:
I also think it's important for CreateInstance to fall back to something closer to the older logic (without depending on ctor order) if IServiceProviderIsService would cause no constructor to be matched.

When we're activating a type the customer does not control without `[ActivatorUtilitiesConstructor]` like our built-in middleware, we don't want that to start breaking code that worked before just because someone is using a third-party container or mocked an incomplete `IServiceProviderIsService` in their test.

Ideally, CreateInstance would only consult IServiceProviderIsService in situations where CreateFactory would always fail and behave the same otherwise.

Reply:
> I also think it's important for CreateInstance to fall back to something closer to the older logic (without depending on ctor order) if IServiceProviderIsService would cause no constructor to be matched.

Can you describe the rules for "something closer to the older logic (without depending on ctor order)"? 

The older logic, AFAICT, picks the constructor that:

1. matched all the provided parameters
2. has a matching parameter in the exact same position in its argument list
3. and is highest matching position (furthest to the end of the parameter list)
4. If multiple ctors meet that criteria, the first was chosen.

The logic came in https://github.com/aspnet/DependencyInjection/pull/14, but there are no comments that describe the intended behavior / rules.